main.nf

```diff
@@ -281,7 +283,7 @@ process BWA {
     
     output:
         file "${sample_id}.log"
-        tuple val(sample_id), path(bam), 
+        tuple val('bwa'), val(sample_id), path(bam), 
               val(align_dir), emit: 'align'
```

Comment:
I am not exactly sure what you want here but I am assuming it is the passing of align_dir?

Reply:
There are multiple issues with this implementation:

1. indexes may be overwritten by other samples creating potential for race conditions and/or the wrong index used for read mapping to a sample
2. if you are going to `publishDir` any files, you need to ensure that the files are uniquely named
3. do not reference published files. Use files/paths passed in explicitly as input via channels. 
4. You shouldn't be creating a `./index` directory. I actually have no idea where it would be saving the files. 


I would rewrite this process like so:


```nextflow
process BOWTIE2 {
  tag "$sample_id vs $ref"
  // don't copy large files like BAM files
  // default is a symlink to work dir BAM file
  publishDir "${params.outdir}/align/bowtie2",
              pattern: "*.bam"
  // easier to just use saveAs to specify the output filename instead of doing it in the script
  publishDir "${params.outdir}/logs/bowtie2",
              pattern: ".command.log",
              saveAs: "bowtie2-${sample_id}-VS-${ref_name}.log"
              mode: "copy"

  input:
  // combine input reads channel with reference FASTAs channel
  tuple sample_id, 
        path(r1), 
        path(r2),
        path(ref_fasta)

  output:
  // sample_id should be first in a tuple so that output channels can be merged 
  // or grouped by sample_id easily
  tuple val(sample_id),
        path(ref_fasta) 
        val('bowtie2'),
        path(bam), emit: 'align'
  // you can just reference .command.log to allow publishing of the log file
  path('.command.log')
  // removed 'file "${sample_id}_bowtie2_align_pe.bam"' since its already captured in the output tuple

  script:
  // need to capture the reference being mapped to
  ref_name = file(ref_fasta).getBasename()
  bam = "bowtie2-${sample_id}-VS-${ref_name}.bam"
  // although it would be more efficient to allow multiple samples to share the 
  // same index by creating a process that outputs the reference index and 
  // using the combine channel operator to outer join the samples with the 
  // indexes, it's alright if this bit of work is repeated for each sample 
  // since our references will typically be quite small so the index building
  // will not take very long compared to read mapping
  """
  bowtie2-build $ref index/$ref_name
  bowtie2 --threads ${task.cpus} \\
      -x index/$ref_name -1 $r2 -2 $r2 \\
  | samtools sort -@${task.cpus} \\
  | samtools view -F4 -b -o $bam;
  """
}
```